File: leet_LIST_141_has_cycle.cpp
/* ******************************************** */
/*                                              */
/*              \\             /`/``            */
/*              ~\o o_       0 0\               */
/*              / \__)      (u  ); _  _         */
/*       / \/ \/  /           \  \/ \/ \        */
/*      /(   . . )            (         )\      */
/*     /  \_____/              \_______/  \     */
/*         []  []               [[] [[]    *.   */
/*         []] []]              [[] [[]         */
/*                                              */
/* ************************************ nuo *** */

//  hare and tortoise method cf. #876 middle of list

class Solution {
public:
    bool hasCycle(ListNode *head)
    {
        ListNode    *H, *T;

        if (!head || !head->next)
            return false ;
        H = T = head;
        while (H->next && H->next->next)
        {
            H = H->next->next;
            T = T->next;
            if (H == T)
                return true ;
        }
        return false ;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

File: leet_LIST_141_has_cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False
        h, t = head, head
        while h.next and h.next.next:
            h = h.next.next
            t = t.next
            if h is t:
                return True
        return False
